File: objc/querySample/querySample/ViewController.m
//
//  ViewController.m
//  querySample
//
//  Created on 2020/12/04.
//

#import "ViewController.h"

@interface ViewController ()

@end

@implementation ViewController

#pragma mark -- dbLib code --

-(void) startConnect {
    
    _db = [[dbLib alloc] init];
    // _db.url = @"http://192.168.0.15/";    // <== change here.
    _db.url = @"http://localhost:8000";    // <== change here.
    _db.verbose = TRUE;
    _db.delegate = self;
    
    dbLibStatus rc = [_db connect];
    
    if(rc != Normal) {
        
        NSLog(@"connect failed %d",rc);
        return;
    }
    
}

-(void) startQuery {
    
    bool rc = [_db query:@"select EMPNO,ENAME from EMP order by EMPNO"];
    
    if(rc){
        NSLog(@"query request failed");
    }
}

-(void) requestCompleted:(NSString*) methodName {
    
    NSLog(@"request completed %@",methodName);
    
    if([methodName isEqualToString:@"Login"]){
        
        [self startQuery];
        
    }
}

-(void) requestFailed:(NSString*)methodName :(NSError *)error {
    
    NSLog(@"request failed %@ %@",methodName,_db.lastErrorText);
        
}

-(bool) rowFetched:(NSMutableDictionary *)row {
    
    NSString* empno = [row objectForKey:@"EMPNO"];
    NSString* ename = [row objectForKey:@"ENAME"];
    
    NSString* str = [NSString stringWithFormat:@"%@ %@", empno, ename];
    [_array addObject:str];
    
    _count++;
    
    return true;
}

-(void) queryCompleted {
    
    dispatch_async(dispatch_get_main_queue(), ^{
        [self->_db disconnect];
        [self->_tableView reloadData];
    });
    
}


- (void)viewDidLoad {
    [super viewDidLoad];
    _count = 0;
    _array = [NSMutableArray array];
    
    [self startConnect];

}

#pragma mark -- table view delegate --

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}


// Customize the number of rows in the table view.
- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return _count;
}


// Customize the appearance of table view cells.
- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    
    static NSString *CellIdentifier = @"Cell";
    
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:CellIdentifier];
    if (cell == nil) {
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:CellIdentifier];
    }
    cell.textLabel.text = [_array objectAtIndex: indexPath.row];
    
    return cell;
}


- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
}



@end

File: swift/querySample/querySample/ViewController.swift
//
//  ViewController.swift
//  querySample
//
//  Created on 2020/12/05.
//

import UIKit

class ViewController: UIViewController,UITableViewDelegate,UITableViewDataSource, dbLibDelegate {
    @IBOutlet weak var _tableView: UITableView!
    private var _rows:NSMutableArray = NSMutableArray.init()
    private var _db:dbLib!
    
    //
    //  MARK: dbLib codes
    //
    func startConnect() {
        _db = dbLib.init()
        _db.url = "http://192.168.0.18:8000"    // <== change here.
        // _db.url = "http://192.168.179.8/"    // <== change here.
        _db.verbose = true
        _db.delegate = self
        
        let rc = _db.connect()
        
        if(rc != Normal) {
            print("connect failed \(rc)")
            return;
        }
    }
    
    func startQuery() {
        let rc = _db.query("select EMPNO,ENAME from EMP order by EMPNO")
        if rc != Normal {
            print("query request failed \(rc)");
        }
    }
    
    func rowFetched(_ row: NSMutableDictionary!) -> Bool {
        let empno:String = row.value(forKey: "EMPNO") as! String
        let ename:String = row.value(forKey: "ENAME") as! String
        let line = empno + " " + ename
        _rows.add(line)
        return true
    }
    
    func requestCompleted(_ methodName: String!) {
        
        print("request completed \(String(describing: methodName))")
        if methodName == "Login" {
            self.startQuery()
        }
        
    }
    
    func requestFailed(_ methodName: String!, _ error: Error!) {
        print("request failed \(String(describing: methodName)) \(String(describing: error))")
    }
    
    func queryCompleted() {
        DispatchQueue.main.async {
            print("queryCompleted")
            self._tableView.reloadData()
        }
    }
    
    //
    //  MARK: view interactions
    //
    
    override func viewDidLoad() {
        super.viewDidLoad()
        _tableView.register(UITableViewCell.self, forCellReuseIdentifier: "Cell")
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        self.startConnect()
    }
    
    //
    //  MARK: table view
    //
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return _rows.count
    }

    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell:UITableViewCell = tableView.dequeueReusableCell(withIdentifier: "Cell")!
        let line = _rows[indexPath.row] as! String
        cell.textLabel!.text = line
        return cell
    }

    func tableView(_ tableView: UITableView, willDisplay cell: UITableViewCell, forRowAt indexPath: IndexPath) {

        tableView.separatorInset = UIEdgeInsets.zero
        tableView.layoutMargins = UIEdgeInsets.zero
        cell.layoutMargins = UIEdgeInsets.zero

    }

    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)
    }

}

